doc/arch/adr-004-transition-mainstream-to-publishing-api-index.md

```diff
@@ -0,0 +1,107 @@
+This is something we need to do after adding new fields or otherwise changing
+the Elasticsearch mappings, for those fields to work properly.
+
+Note that neither of these mechanisms add or remove documents from the search index.
+
+- If an edition is ever published, and the search index doesn't get updated, it stays out of date until the next time the document is updated
+
+- If an edition is unpublished without Rummager being notified, it stays in the search index forever
+
+Long term, we'd like to be able to easily rebuild the whole `govuk` index
+from publishing API, to avoid these kind of problems, but we aren't planning to do this right now.
+
+### Immediate needs
+
+We need to be able to populate the `govuk` index with all the documents that
+already exist, so we can start using this index in the Search API.
+
+We also want to be able to reindex everything that has been published within a
+short period of time (1 or 2 days), so that we can easily recover the index
+from backups.
+
+### Moving formats one at a time
+We're intending to switch on the new index format-by-format, by querying both old
+and new indexes, and using filters to select which index gets used for each
+format.
+
+This means we can retire old search indexing code in publishing apps, and search indexes that are no longer needed, without having to populate *everything* all at once.
+
+We can also revert back to the old index with a simple configuration change if something goes wrong.
+
+A problem we've discovered with this approach is that the relevancy of a document within an index depends on the other documents in that index. If an index is only partially populated, the [TF-IDF statistics](https://www.elastic.co/guide/en/elasticsearch/guide/current/scoring-theory.html#tfidf) are less representative, and this affects search results.
+
+## Decision
+Firstly, we'll implement a task to bulk-reindex chunks of content from publishing API. Rummager will process this content in the same way as regular publishing updates.
+
+- Reindexing by format lets us initially populate the new index.
+
+- Reindexing by date range lets us bring the `govuk` index up to date when restoring from a backup.
+
+Secondly, when we change the indexing process for a format to use the new index, the format will go through the following phases:
+
+### Phase 1: Untransitioned
+At search time, Rummager reads untransitioned formats from the old indexes. Documents belonging to untransitioned formats that are stored in the `govuk` index will be filtered out.
+
+At index time, Rummager will ignore publishing API messages affecting untransitioned formats.
+
+The nightly update job will update the popularity field of untransitioned formats in the `mainstream` index. Then it will copy untransitioned format documents from the `mainstream` index into the `govuk` index.
```

Comment:
This step is not required as we won't be using the popularity field for untransitioned documents. The popularity is a boost after the TF-IDF calculation. :)

Reply:
I'm not sure what you mean here - this is talking about docs that get returned from the mainstream index, so they still get their popularity field updated there.

Reply:
What I mean is that they don't need to have a valid popularity. This is because they will never be scored as they will never be returned. So we could easily do the copy to govuk before/during the mainstream popularity update as after it is finished.

Reply:
Ah I see. I'll clarify this